java/vector/src/main/java/org/apache/arrow/vector/compression/CompressionCodec.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.compression;
+
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+
+/**
+ * The codec for compression/decompression.
+ */
+public interface CompressionCodec {
+
+  /**
+   * Compress a buffer.
+   * @param allocator the allocator for allocating memory for compressed buffer.
+   * @param unCompressedBuffer the buffer to compress.
+   *                           Implementation of this method should take care of releasing this buffer.
+   * @return the compressed buffer.
+   */
+  ArrowBuf compress(BufferAllocator allocator, ArrowBuf unCompressedBuffer);
```

Comment:
In thinking through this interface, something occurred to me.  Given the way the current Message serialization code works is it forces a copy from an underlying channel.  Given this, it seem like we would now have two copy like operations:
1.  Copy from channel to compressed buffer
2.  Uncompress the buffer when loading VectorSchemaRoot.

I was wondering what you thought about deferring compression until writing out the  batch and doing decompression eagerly when reading in the batch to avoid one of these copies.  I think one thing that would potentially help us decide on the course to take would be look at the possible libraries we would use for compression/decompression and whether they support off-heap, InputStream or some other interface for accomplishing the compression

@liyafan82 have you done more research in this area?

@rymurr do you have anything thoughts on this?

Reply:
@emkornfield Thank you for starting this discussion and sharing your good ideas. 
Your reasoning makes sense to me. 

I guess I was looking at the problem from a different perspective. 

IMO, the bottleneck of a compressing codec is the CPU resource, and the main purpose of compressing is to reduce memory/network bandwidth consumption.

Given the above assumptions, we should try to do the compression as early as possible. The earliest possible place should be in the `getFieldBuffers` method. In this PR, we do it in `VectorUnLoader`, which is not the best, but close enough to the best. Similarly, we should try to do the decompression as late as possible. In this PR, we do it in `VectorLoader`, which is close to the optimal.

Admittedly, we have additional copies after introducing the compression framework. However, both additional copies are based on the compressed data, with reduced data size, so the overhead should be small.

The above reasoning is based on the assumption that the compression codec could effectively reduce the data size, which is not always true in practice. So I think we can make the decision based on the specific compression codec, and real benchmark data?

Reply:
@liyafan82 good points.  I think the other use-case I didn't consider is Flight which I think will already be in off-heap memory. 

In terms of copies, the should be smaller in general, but it would be good to to understand if there are any limitations.  It looks like at least [airlift supports off-heap decompression](https://github.com/airlift/aircompressor/blob/master/src/main/java/io/airlift/compress/lz4/Lz4RawDecompressor.java), so I think this API is probably OK from that perspective.

@rymurr did you have any other comments otherwise I can take another look and then I think we can merge.

Reply:
@emkornfield Do you think we can merge this now?

Reply:
yes, I think so.  i'll merge